pkg/querier/worker/scheduler_processor.go

```diff
@@ -304,6 +320,61 @@ func (sp *schedulerProcessor) runRequest(ctx context.Context, logger log.Logger,
 	}
 }
 
+func consumeResponseStreamingEnabledHeader(headers []*httpgrpc.Header) ([]*httpgrpc.Header, bool) {
+	streamEnabledViaHeader := false
+	for i, header := range headers {
+		if header.Key == ResponseStreamingEnabledHeader {
+			if header.Values[0] == "true" {
+				streamEnabledViaHeader = true
+			}
+			headers = append(headers[:i], headers[i+1:]...)
+			break
+		}
+	}
+	return headers, streamEnabledViaHeader
+}
+
+func streamResponse(ctx context.Context, c client.PoolClient, queryID uint64, response *httpgrpc.HTTPResponse, stats *querier_stats.Stats) error {
+	sc, err := c.(frontendv2pb.FrontendForQuerierClient).QueryResultStream(ctx)
+	if err != nil {
+		return fmt.Errorf("error creating stream to frontend: %w", err)
+	}
+
+	// Send metadata
+	err = sc.Send(&frontendv2pb.QueryResultStreamRequest{
+		QueryID: queryID,
+		Data: &frontendv2pb.QueryResultStreamRequest_Metadata{Metadata: &frontendv2pb.QueryResultMetadata{
+			Code:    response.Code,
+			Headers: response.Headers,
+			Stats:   stats,
+		}},
+	})
+	if err != nil {
+		return fmt.Errorf("error sending initial response to frontend: %w", err)
+	}
+
+	// Send body chunks.
+	for offset := 0; offset < len(response.Body); {
+		err = sc.Send(&frontendv2pb.QueryResultStreamRequest{
+			QueryID: queryID,
+			Data: &frontendv2pb.QueryResultStreamRequest_Body{Body: &frontendv2pb.QueryResultBody{
+				Chunk: response.Body[offset:min(offset+responseStreamingBodyChunkSizeBytes, len(response.Body))],
```

Comment:
I'm not sure, but intuitively it makes sense to avoid the extra overhead of two rpc calls for response bodies that fit into a single chunk. I have added this in 1bb0df7ec.